fabtests/functional/multi_ep.c

```diff
@@ -245,6 +274,55 @@ static int do_transfers(void)
 			return ret;
 	}
 
+	for (i = 0; i < num_eps; i++) {
+		remote_keys[i] = ((uint64_t *)recv_bufs[i])[0];
+		remote_addrs[i] = ((uint64_t *)recv_bufs[i])[1];
+	}
+
+	printf("PASSED multi ep sends\n");
+	return 0;
+}
+
+static int do_rma(void)
+{
+	int i, ret, cq_read_idx;
+	uint64_t cur;
+
+	for (i = 0; i < num_eps; i++)
+		recv_bufs[i][0] = 'A';
+
+	ft_sync();
+
+	for (i = 0; i < num_eps; i++) {
+		ret = ep_post_write(i);
+		if (ret) {
+			FT_PRINTERR("fi_write", ret);
+			return ret;
+		}
+	}
+
+	printf("Wait for all messages from peer\n");
+	for (i = 0; i < num_eps; i++) {
+		if (shared_cq)
+			cq_read_idx = 0;
+		else
+			cq_read_idx = i;
+		cur = 0;
+		ret = ft_get_cq_comp(txcqs[cq_read_idx], &cur, 1, -1);
+		if (ret < 0)
+			return ret;
+	}
+
+	//sync to make sure data is there
+	for (i = 0; i < num_eps; i++) {
+		if (shared_cq)
+			cq_read_idx = 0;
+		else
+			cq_read_idx = i;
+		while (recv_bufs[i][0] == 'A')
```

Comment:
For EFA, I think this wait strategy will fail unless NIC deliver the data in the buffer in sequence. We have some problem for it before and currently we only use this method when NIC supports RDMA write with in-order delivery. Why not use fi_writedata here so you can rely on the CQ poll to know data is delivered?

Reply:
Yeah that makes sense, I can change it. Thanks!